hive-metastore/src/main/java/org/apache/iceberg/hive/legacy/LegacyHiveSchemaUtils.java

```diff
@@ -0,0 +1,691 @@
+
+      if (isSchemaEvolved(oldField.schema(), newField.schema())) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private static boolean isSchemaEvolved(Schema oldSchema, Schema newSchema) {
+    if (oldSchema == null || newSchema == null) {
+      throw new IllegalArgumentException("The input schemas cannot be null");
+    }
+
+    if (!oldSchema.getType().equals(newSchema.getType())) {
+      if (oldSchema.getType().equals(Schema.Type.ENUM) && newSchema.getType().equals(Schema.Type.STRING)) {
+        return false;
+      }
+
+      LOG.info("Schema is evolved. The old schema type is: {}, the new schema type is: {}",
+          oldSchema.getType(), newSchema.getType());
+      return true;
+    }
+
+    switch (newSchema.getType()) {
+      case BOOLEAN:
+      case BYTES:
+      case DOUBLE:
+      case FLOAT:
+      case INT:
+      case LONG:
+      case STRING:
+      case FIXED:
+        return false;
+      case ENUM:
+        return oldSchema.getEnumSymbols().size() != newSchema.getEnumSymbols().size();
+
+      case RECORD:
+        return isRecordSchemaEvolved(oldSchema, newSchema);
+
+      case MAP:
+        return isSchemaEvolved(oldSchema.getValueType(), newSchema.getValueType());
+
+      case ARRAY:
+        return isSchemaEvolved(oldSchema.getElementType(), newSchema.getElementType());
+
+      case UNION:
+        boolean isBothNullableType =
+            AvroSchemaUtil.isOptionSchema(oldSchema) && AvroSchemaUtil.isOptionSchema(newSchema);
+
+        if (!isBothNullableType) {
+          throw new IllegalArgumentException(
+              "We do not support UNION type " + "except nullable field in schema: " + newSchema.toString(true));
+        }
+
+        Schema oldNonNullType = AvroSchemaUtil.fromOption(oldSchema);
+        Schema newNonNullType = AvroSchemaUtil.fromOption(newSchema);
+
+        return isSchemaEvolved(oldNonNullType, newNonNullType);
+      default:
+        throw new IllegalArgumentException(
+            "Unsupported Avro type " + newSchema.getType() + " in new schema: " + newSchema.toString(true));
+    }
+  }
+
+  private static boolean isPartitioned(@Nonnull Table tableOrView) {
+    Preconditions.checkNotNull(tableOrView, "tableOrView cannot be null");
+
+    List<FieldSchema> partitionColumns = getPartitionCols(tableOrView);
+
+    if (partitionColumns == null) {
+      return false;
+    }
+
+    return partitionColumns.size() != 0;
+  }
+
+  private static List<FieldSchema> getPartitionCols(@Nonnull Table tableOrView) {
+    Preconditions.checkNotNull(tableOrView, "tableOrView cannot be null");
+
+    List<FieldSchema> partKeys = tableOrView.getPartitionKeys();
+    if (partKeys == null) {
+      partKeys = new ArrayList<>();
+      tableOrView.setPartitionKeys(partKeys);
+    }
+
+    return partKeys;
+  }
+
+  private static void mergeAndAppendField(Schema.Field sourceField, Schema.Field metaDataField,
```

Comment:
`applySupplementalFieldToHiveField()`. Also keep the order of the parameters consistent.